rpc/gnmi/gnmi-union_replace.md

```diff
@@ -0,0 +1,327 @@
+
+A gNMI `SetRequest` `replace` operation using origin CLI must use only one  `replace` operation which contains the entire contents of the desired CLI configuration.  One or more `update` operations may follow the `replace` operation. The contents of the `update` are appended to the contents of that specified in the `replace` operation.
+
+Note, in a gNMI SetRequest the `delete`, `replace` and `update` operations are performed in order as prescribed by the gNMI specification.  In origin CLI, the delete operation is unsupported.
+
+### 4.2 Replace operation for path based Native Origin behavior
+
+A `replace` operation for path based native origins should be performed in the same way as OpenConfig as defined in gNMI Specification section 3.4.
+
+## 5 Device behavior for mixed native/CLI OC origin  SetRequests
+
+### 5.1 Objective
+
+For this use case, the goal is to use OC wherever possible and a native origin only where OC support is lacking.  The configuration generation software must understand the device specific impact of native origin commands and their interaction with OC, if any.
+
+### 5.2 Requirements for SetRequest with 2 origins
+
+Requirements are defined where 2 origins are present in a SetRequest.
+Note, in the below requirements, the use of “SetRequest operation” refers to the type of operations used in a “SetRequest” message.
+
+#### 5.2.2 union_replace operation containing a native origin and origin OpenConfig
+
+A new operation named `union_replace` is introduced to meet the “full configuration replace” goal using exactly one of NY or CLI, along with OC.
+
+In this scenario, the target should replace the contents of its underlying configuration using the union of the contents of the `union_replace` operation.  Union means the target must join the contents (e.g., the native and OpenConfig) together to form a single configuration change which replaces the entire configuration of the target.  If the target cannot join the contents together or replace the configuration, the target MUST return an error.
+
+An abstract set of steps a target may use to achieve `union_replace` behavior are:
+
+1. If using CLI origin, create a candidate configuration using factory defaults within the applicable namespace.  Factory defaults means a configuration with only the default values set by the NOS.  For example a NOS factory default configuration may assert that all ports, except a single management port are shutdown. The NOS may also assert that a DHCP client is enabled on the single management port and that protocols like BGP and ISIS are not instantiated and therefore are also not enabled).  If using another native origin, create a candidate configuration using the current running configuration.
+
+1. Union the native data with the OC paths data and apply this to the candidate  configuration.  The union behavior is described in 5.3 Union behavior options.
```

Comment:
I think more clarification is still needed on this.

If in step 1 the intention is that there the possibility of explicit config existing in the running datastore (i.e. "show configuration" returns something, not just a completely blank/empty response), and that config includes statements that are deletable/removable by a client (e.g. a user called "admin"), then the new step 2 proposed would delete all those deletable lines of config. In that case I'm not sure the point of step 1. Anything would just be removed by step 2.

But if the intention after step 1 is that there is no explicit config (i.e. "show configuration" returns absolutely nothing), just an empty config where the NE is using some default values, then we should clarify that in step 1.

Reply:
If a line of config is deletable but not default, why is it returned by 1? As I read it, anything that is a default, whether returned by some legacy CLI method ;-) or not should be in the candidate that is to be applied.

Reply:
I'm not positive what's being described in #1 wrt "factory default", but worried that term may cause confusion since it is a standard term in the IETF. Factory default config in the IETF looks & feels like a small set of config that a user/client explicitly added.  In other words, a gNMI Get would return some data (e.g. user "admin", etc).

But that's very different from default values in a schema. For example: let's pretend an interface mtu has a default of 1500 in a YANG schema. If a client never sets that mtu leaf, then reading back the config with a Get would not return any leaf called mtu. But the system would be using an mtu of 1500 down in the app layer. IMO I don't consider the mtu leaf to be in the configuration/candidate (i.e. there is nothing there in the instance data for the mtu).

I'm not clear if #1 is saying a Get would return:
(A) absolutely no data (i.e. default values are being used in the app layer, but there is no config at all in the running DS), or 
(B) the Get would return some data (i.e. like the IETF "factory default" scenario, where maybe a user "admin" might be returned in the get response).  

From what I can tell it must be (A). But then why don't we just say the following for #1:
1. Start with an empty candidate configuration (no instance data, a Get would return an empty data set)

Reply:
I'm similarly unsure about the last sentence of bullet 1: _If using another native origin, create a candidate configuration using the current running configuration._

If step 2 replaces the config, then why are we starting with the current running vs an empty candidate? Maybe we'd want to mention that either is equivalent (to clarify there isn't something subtle/special here about starting with the current running).

Reply:
I will add a definition for factory default configuration.  It will read something like:
"The configuration that is present on the device when it is powered on for the first time by the user."

When using union_replace, the user is expecting to assert all configuration items needed to ensure the device has the intended state.    If a device factory default configuration includes an admin account and the user wants this removed, the user must include the negation of that user in the union_replace.  

gNMI operations, including union_replace, are expected to be idempotent.  So each and every union_replace operation would include this negation, but it would only cause a change in the 'running' configuration on the first SetRequest.  Subsequent SetRequest would include the negation, but there would be no effect on the 'running' configuration or 'startup' configuration because that admin account was already removed.  

Regarding  native yang (NY) origin the user can choose to replace from the root NY tree or at subtrees.  In CLI, this is not feasible because there is no path structure, so we chose to require a 'factory default' candidate configuration.

Reply:
The negation of an admin account (in a declarative config) is simply the absence of it in the combined native+OC union config that was sent in the union_replace. When you say "must include the negation" it makes it sound like you want the instruction "no user admin" somehow in the union_replace (i.e. an imperative statement from the "old days" vs a declarative configuration).